broker/pkg/coordinator/coordinator.go

```diff
@@ -52,6 +62,24 @@ func (c Consumer) GetTopic() []byte {
 	return []byte(c.Topic)
 }
 
+func (c *Consumer) SetStatus(status protocol.ConsumerStatus) {
+	c.mu.Lock()
+	c.status = status
+	c.mu.Unlock()
+}
+
+func (c *Consumer) LoadAvailableBufferSize() uint32 {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	return c.status.AvailableBufferSize
+}
+
+func (c *Consumer) AddAvailableBufferSize(delta uint32) {
```

Comment:
I think is not named correctly "buffer", it seems more like this is a "sendMeXMessages" request, rather than a buffer.  And I would not use a delta, but rather fixed values. I don't have a solution in mind right now, but the concern is that the Client will have a hard time adjusting the values, eg:
Client: send me 20 msgs
Server: here is 5 (not enough)
Client: send me 20
server here is 35 

Either way this is a tricky system

Reply:
The consumer updates it's status (including AvailableBufferSize) using SetStatus  which is not a delta but a replacement of the old value with a new one. 

AvailableBufferSize is the actual the number of messages the consumer has available in its buffer.

AddAvailableBufferSize is used by the Broker to update the AvailableBufferSize after it sends a number of messages to the consumer so it can keep track of how many messages it can send until the buffer if full. The consumer is responsible to update it's status, implicit AvailableBufferSize value, periodically or after it processes some messages, if the consumer does not update this value the broker will sens messages until the AvailableBufferSize is 0.